### DISCUSSION TITLE
help request:  [file-logger] duplicate log, new log file contains history log message

### DISCUSSION THREAD

**Discussion by kizy008** [apache/apisix#7902] [Q&A] [answered]
### Description

In my project, I need split log file by hour and delete logs 7 days ago. So I write a rotate plugin like [log-rotate.lua], and it works.
But I found that every new log file contains the history log message. For example, [log-2022-09-01_01-00-00.log] contains all message in [log-2022-09-01_00-00-00.log]]. So every log contains all history message.  I think the new log should contains message in its hour period. But I got all history message. 
I want to know how to get incremental logs in new log file.

我的项目中需要将日志按照小时分割并自动淘汰7天前的历史日志。 所以我仿照[log-rotate.lua]写了file-logger的滚动切割日志插件，并且按照预期生效了。但是我发现，所有的心日志文件都包含了全量的历史日志。比如 [log-2022-09-01_01-00-00.log]日志不仅包含了23点到0点的日志， 还包含了所有的历史全量日志。我本以为新日志应该只包含过去1个小时之内的日志。
我想知道如何利用file-logger在每个日志文件中保存这个小时时间段内增量的日志呢？

### Environment

- APISIX version (run `apisix version`):
- Operating system (run `uname -a`):
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

**Comment by AlphaChen** [comment]
i meet same problem。can u show u code？

**Comment by juststillthinking** [comment]
1. rename old log file and create new one
2. notify other processes to reopen the file

Do you follow the two steps ?

**Comment by AlphaChen** [comment]
> 1. rename old log file and create new one
> 2. notify other processes to reopen the file
> 
> Do you follow the two steps ?

how to notify other processes

**Comment by juststillthinking** [comment, marked as answer]
You could use [ngx.shared.dict](https://github.com/openresty/lua-nginx-module#ngxshareddict) to notify other processes.
For example

```
-- process 1
local rotate = ngx.shared.rotate
rotate:set("reopen", true)
```

```
-- process 2
local rotate = ngx.shared.rotate
local reopen = rotate:get("reopen")
if reopen then
    -- reopen file
end
```

**Reply by kizy008** [reply]
`local function rotate()
    local interval = defaults.interval
    local max_kept = defaults.max_kept
    local logPath  = defaults.log_path
    local logName  = defaults.log_name

    core.log.info(plugin_name, " interval: ", interval)
    core.log.info(plugin_name, " max_kept: ", max_kept)
    core.log.info(plugin_name, " logPath: ", logPath)
    core.log.info(plugin_name, " logName: ", logName)

    if not default_logs then
        default_logs = {}
        init_default_logs(default_logs, logPath, logName)
    end

    ngx_update_time()
    local now_time = ngx_time()
    if not rotate_time then
        -- first init rotate time
        rotate_time = now_time + interval - (now_time % interval)
        core.log.info("first init file-logger-rotate rotate time is: ", rotate_time)
        return
    end

    if now_time < rotate_time then
        -- did not reach the rotate time
        core.log.info("file-logger-rotate rotate time: ", rotate_time, " now time: ", now_time)
        return
    end

    local now_date = os_date("%Y-%m-%d_%H-%M-%S", now_time)
    if false == file_exists(default_logs[logName].file) then
        local log_file = file_create(default_logs[logName].file)
        if not log_file then
            core.log.warn("file-logger-rotate [", default_logs[logName].file, "] not exist and create failed, no need rotate")
            rotate_time = rotate_time + interval
            return
        end
    end
    local log_new_file = rename_file(default_logs[logName], now_date)
    if not log_new_file then
        -- reset rotate time
        rotate_time = rotate_time + interval
        return
    end
    -- create log file after rename
    local log_file = file_create(default_logs[logName].file)
    if not log_file then
        core.log.warn(plugin_name, " create new log file failed")
    end
    core.log.warn("send USR1 signal to master process [",
            ngx_process.get_master_pid(), "] for reopening log file")
    local ok, err = signal.kill(ngx_process.get_master_pid(), signal.signum("USR1"))
    if not ok then
        core.log.error("failed to send USR1 signal for reopening log file: ", err)
    end

    -- clean the oldest file
    local log_list, log_dir = scan_log_folder(logPath, logName)
    for i = max_kept+1, #log_list.log_file do
        local path = log_dir .. log_list.log_file[i]
        ok, err = os_remove(path)
        if err then
            core.log.error("remove old file-logger-rotate file log: ", path, " err: ", err, " res: ", ok)
        end
    end

    -- reset rotate time
    rotate_time = rotate_time + interval
end`

here is my rotate function. After rename the log file, I send USR1 singal to nginx. But the new log remain history log message.

**Reply by kizy008** [reply]
I will try shared_dict to notify file-logger

**Reply by juststillthinking** [reply]
The log file in file-logger is not managed by nginx, so send signal to nginx master will not help.

**Reply by kizy008** [reply]
I change my code here:
rotate-plugins:
`local function rotate()
    local interval = defaults.interval
    local max_kept = defaults.max_kept
    local logPath  = defaults.log_path
    local logName  = defaults.log_name

    core.log.info(plugin_name, " interval: ", interval)
    core.log.info(plugin_name, " max_kept: ", max_kept)
    core.log.info(plugin_name, " logPath: ", logPath)
    core.log.info(plugin_name, " logName: ", logName)

    if not default_logs then
        default_logs = {}
        init_default_logs(default_logs, logPath, logName)
    end

    ngx_update_time()
    local now_time = ngx_time()
    if not rotate_time then
        -- first init rotate time
        rotate_time = now_time + interval - (now_time % interval)
        core.log.info(plugin_name, "first init file-logger-rotate rotate time is: ", rotate_time)
        return
    end

    if now_time < rotate_time then
        -- did not reach the rotate time
        core.log.info(plugin_name, "file-logger-rotate rotate time: ", rotate_time, " now time: ", now_time)
        return
    end

    local now_date = os_date("%Y-%m-%d_%H-%M-%S", now_time)
    if false == file_exists(default_logs[logName].file) then
        local log_file = file_create(default_logs[logName].file)
        if not log_file then
            core.log.warn(plugin_name,"file-logger-rotate [", default_logs[logName].file, "] not exist and create failed, no need rotate")
            rotate_time = rotate_time + interval
            return
        else
            core.log.warn(plugin_name, "file-logger-rotate [", default_logs[logName].file,"] create success")
        end
    end
    local log_new_file = rename_file(default_logs[logName], now_date)
    if not log_new_file then
        -- reset rotate time
        core.log.error(plugin_name,"rename file got empty new file name reset rotate time here and return immediately")
        rotate_time = rotate_time + interval
        return
    else
        core.log.warn(plugin_name, "rename file [", default_logs[logName].file, "] to [", log_new_file, "] success")
    end
    if true == file_exists(default_logs[logName].file) then
        core.log.error(plugin_name, "file [", default_logs[logName].file,"] exist after rename, remove it")
        local ok, err = os_remove( default_logs[logName].file)
        core.log.error(plugin_name,  "remove [", default_logs[logName].file, "] after rename ok: ", ok, "err: ", err)
    else
        core.log.warn(plugin_name, "file [", default_logs[logName].file,"] not exist after rename")
    end

    -- set reopen in shared dict, expire time 5s
    local rotate_shdict = ngx.shared.file_log_rotate
    local succ, err, forcible = rotate_shdict:set("reopen", true, 5)
    core.log.warn(plugin_name, "save reopen flat to  ngx.shared.file_log_rotate, expire time 5, succ = ", succ, "err = ", err, "forcible = ", forcible)

    core.log.warn("send USR1 signal to master process [",
            ngx_process.get_master_pid(), "] for reopening log file")
    local ok, err = signal.kill(ngx_process.get_master_pid(), signal.signum("USR1"))
    if not ok then
        core.log.error("failed to send USR1 signal for reopening log file: ", err)
    end
    -- create log file after rename
    local log_file = file_create(default_logs[logName].file)
    if not log_file then
        core.log.error(plugin_name, " create new log file failed")
    else
        core.log.warn(plugin_name, "create new log file [", default_logs[logName].file, "] success")
    end


    -- clean the oldest file
    local log_list, log_dir = scan_log_folder(logPath, logName)
    for i = max_kept+1, #log_list.log_file do
        local path = log_dir .. log_list.log_file[i]
        ok, err = os_remove(path)
        if err then
            core.log.error("remove old file-logger-rotate file log: ", path, " err: ", err, " res: ", ok)
        end
    end

    -- reset rotate time
    rotate_time = rotate_time + interval
end`

And file-logger here:
`local open_file_cache
if is_apisix_or then
    -- TODO: switch to a cache which supports inactive time,
    -- so that unused files would not be cached
    local path_to_file = core.lrucache.new({
        type = "plugin",
    })

    local function open_file_handler(conf, handler)
        local file, err = io_open(conf.path, 'a+')
        if not file then
            return nil, err
        end

        handler.file = file
        handler.open_time = ngx.now() * 1000
        return handler
    end

    function open_file_cache(conf)
        local last_reopen_time = process.get_last_reopen_ms()

        local handler, err = path_to_file(conf.path, 0, open_file_handler, conf, {})
        if not handler then
            return nil, err
        end

        if handler.open_time < last_reopen_time then
            core.log.notice("reopen cached log file: ", conf.path)
            handler.file:close()

            local ok, err = open_file_handler(conf, handler)
            if not ok then
                return nil, err
            end
        end

        -- 通过 shared dict 获知其他进程reopen通知
        local rotate_shdict = ngx.shared.file_log_rotate
        local reopen_flag = rotate_shdict:get("reopen")
        if reopen_flag then
            core.log.warn(plugin_name, "receive reopen from shared dict, reopen cached log file: ", conf.path)
            handler.file:close()
            local ok, errInfo = open_file_handler(conf, handler)
            if not ok then
                return nil, errInfo
            end

        end

        return handler.file
    end
end`

And Here is the log output:
![image](https://user-images.githubusercontent.com/12572997/189910957-01cf1b3e-1ec6-463d-8164-92b5f448a959.png)
The Problem still occurs.
![image](https://user-images.githubusercontent.com/12572997/189910855-8c541e0b-75ad-4644-94b5-c0b8978d357d.png)

**Reply by kizy008** [reply]
> The log file in file-logger is not managed by nginx, so send signal to nginx master will not help.

Did you solve this problem by adding shared dict ?

**Reply by kizy008** [reply]
![image](https://user-images.githubusercontent.com/12572997/189916168-5a4cba8f-68db-4b0b-896e-95263ba92112.png)
Here is the process new log file created. 
Firstly the file.log rename success. Then create a new file.log, But the new log contains all history log message. I don't know how it happens.

**Reply by kizy008** [reply]
![image](https://user-images.githubusercontent.com/12572997/190046815-a6c50c5f-9b6e-4aa8-a3a6-742a8aead0f7.png)
![image](https://user-images.githubusercontent.com/12572997/190047077-e875a163-c2e9-4d44-95e4-7b1fc7440e4c.png)

I add some log when create new log file, and I found that new log file have history log message. I don't know why. I open new log file with mode "w".  It should not have any log message.

**Reply by kizy008** [reply]
Maybe it related to storage medium. I use COS (Cloud Object Storage) to store file logs. And I found that when I use hard disk or CBS , the problem did not occur. When I use COS, I rename a new log file with size 0 to former log name, the log file would be no longger empty.

我感觉这个问题可能跟底层存储介质有关。我使用COS存储日志。当我使用硬盘或云硬盘时，没有出现过这种日志累加的问题。我使用COS来存储时， 当我将一个大小为0的空文件重命名为原先的日志名称时，我发现日志文件不再为空，里面包含了所有的历史日志记录。

Perhaps， I will save history log in COS and real time log in hard disk. Then rotate real time log and save them to history log direcotry.
也许，我应该将历史日志存储在COS上，将实时日志存储在硬盘上，然后切割实时日志，将切割后的日志存储在历史日志的目录里面。

**Reply by kizy008** [reply]
This issue has been solved by write real time log on hard disk, and rotate the log then move to history log directory on COS.
There is a step need to be  attentioned. Because COS and hard disk are diffrent, if we use os.rename, it will report [os.rename Invalid cross-device link]. So here I use os.execute instead. 
It works well for me, I hope this solution may help you.

**Comment by tzssangglass** [comment]
Since this is your own code, not related to the project, I turned it into discussion.
